src/main/java/com/palantir/gradle/versions/VersionsLockPlugin.java

```diff
@@ -769,25 +772,49 @@ private static GcvScope getScopeRecursively(
             return cached.get();
         }
 
-        GcvScope gcvScope = component.getDependents().stream()
-                .filter(dep -> !dep.isConstraint())
-                .filter(dep -> dep.getRequested() instanceof ModuleComponentSelector)
-                .map(dependent -> {
-                    ModuleIdentifier requestedModule =
-                            ((ModuleComponentSelector) dependent.getRequested()).getModuleIdentifier();
-                    // If the dependency came from a project, then the requested ModuleIdentifier should be in the
-                    // edgeScopes. Otherwise, recurse until we find a project dependent.
-                    Optional<GcvScope> maybeScope = directDependencyScopes.getScopeFor(requestedModule);
-                    if (dependent.getFrom().getId() instanceof ProjectComponentIdentifier && maybeScope.isPresent()) {
-                        return maybeScope.get();
-                    }
-                    return getScopeRecursively(dependent.getFrom(), scopeCache, directDependencyScopes);
-                })
+        Set<ResolvedDependencyResult> traversedComponents = new HashSet<>();
+        Deque<ResolvedDependencyResult> stack =
+                new ArrayDeque<>(component.getDependents().size());
+        stack.addAll(component.getDependents());
+
+        Set<GcvScope> discoveredScopes = new HashSet<>(2);
+        while (!stack.isEmpty()) {
+            ResolvedDependencyResult dependent = stack.removeFirst();
+            if (dependent.isConstraint()
+                    || !(dependent.getRequested() instanceof ModuleComponentSelector)
+                    || traversedComponents.contains(dependent)) {
+                continue;
+            }
+
+            Optional<GcvScope> cachedValue = Optional.ofNullable(scopeCache.get(dependent.getFrom()));
+            if (cachedValue.isPresent()) {
+                discoveredScopes.add(cachedValue.get());
+                continue;
+            }
```

Comment:
Previously this was checking that a component i.e. a `ResolvedComponentResult` (whose dependents we later recurse through) is in the cache.

That component was previously coming from `dependent.getFrom()` in the recursive call, but now we're discarding that `dependent.getFrom()`, instead dumping its *dependents* into the stack.

However, from the point of view of the _dependents_, the component is actually `dependent.getSelected()`